### PR TITLE
CI: Merge open blocker PRs in all CI workflows + other improvements

### PR DESCRIPTION
**Pull request by mkoeppe**
The new script `.ci/merge-fixes.sh` uses the GitHub CLI to retrieve open PRs with the "blocker" label from sagemath/sage and tries to merge them into the tested branch, ignoring any PRs that give merge failures.

We call this script in all CI workflows. Thus, 
- hotfixes for the linter (such as #36327) or 
- fixes for build problems (such as #36276)
 
will take effect in the CI workflows of all PRs.

This can be seen in the runs of the workflows of the present PR (section "Merge CI fixes from sagemath/sage").

Additionally,
- we fix `debian-trixie`, which failed because a system package is no longer available
- we fix `macos-conda-maximal` by removing some fictitious distros/conda.txt files
- we refactor ci-macos using a new reusable workflow; example runs: 
   - https://github.com/mkoeppe/sage/actions/runs/6308124398
   - https://github.com/sagemath/cysignals/actions/runs/6316714387?pr=185
- we update the ci-macos platforms, including a configuration for the new and problematic Xcode 15.0 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: .ci/merge-fixes.sh
Comment (kliem) [inline .ci/merge-fixes.sh:5]: Interesting approach. There is a couple of thoughts I have about this:
- We would merge any blocker PR. Is there any quality control, e.g. pipeline passed etc. or positively reviewed.
- This would self-reference blockers.
- Unless we merge in all blockers first, we might merge in PRs that depend on blockers, but do not declare that dependency correctly.

Would there still be a sanity check that just tests this PR without merging in other things?

Comment (mkoeppe) [reply .ci/merge-fixes.sh:5]: > * We would merge any blocker PR. Is there any quality control, e.g. pipeline passed etc. or positively reviewed.

There isn't, and I don't think it's needed. Since the transition to GitHub, using priority labels has become pretty rare, and no one has used the blocker tag without a good reason.

Comment (mkoeppe) [reply .ci/merge-fixes.sh:5]: > * This would self-reference blockers.

Yes, but for these the merge would be a no-op, and I think I'm handling this correctly in the scripts. 

EDIT: See the run for the current workflow.

Comment (mkoeppe) [reply .ci/merge-fixes.sh:5]: > * Unless we merge in all blockers first, we might merge in PRs that depend on blockers, but do not declare that dependency correctly.

Currently, the release manager scripts ignore all dependencies and all priorities anyway...

**Comment by tobiasdiez** [comment]
I am not the biggest fan of this auto-apply of blocker PRs in all work flows. I think it would be better to attack the issue at the root and make it impossible to merge PRs with failing workflows. Everything else feels like a weird hack, especially since one still has to create this hotfix PRs in the first place.

**Comment by mkoeppe** [comment]
Yes, this PR is a practical solution. It is designed to avoid two approaches, each of which would be more elegant:
- It does not create hotfix branches explicitly. Hence, there is no hotfix branch to maintain; hence there is not need for a maintainer of the hotfix branch.
- It does not require communication with the release manager.

This careful design makes this PR an effective solution.

File: .ci/merge-fixes.sh
Comment (kliem) [reply .ci/merge-fixes.sh:5]: Sounds plausible.

**Review by kliem — APPROVED** [review]

**Comment by mkoeppe** [comment]
Thanks @kliem 

> One should agree though that the blocker label is only applied once the author confident that this is a solution.

Setting the priority label is limited to the Triage role; I think we can trust that developers who have been elevated to this role will not abuse it.

**Comment by kliem** [comment]
> Thanks @kliem
> 
> > One should agree though that the blocker label is only applied once the author confident that this is a solution.
> 
> Setting the priority label is limited to the Triage role; I think we can trust that developers who have been elevated to this role will not abuse it.

Cool. As you noticed, I'm not very involved at the moment and I wasn't aware that the move to github enabled more granular control of such things.

**Comment by tobiasdiez** [comment]
> The CI is rather fragile depending on a bunch of system packages (and so is the entire installation process). Package updates will break building on certain environments. So I think this is a reasonable solution.

Thats not the problem. The CI doesn't randomly fails. It fails because PRs with broken CI are merged since the release manager takes only a different set of CI checks into account.

A more elegant and effective solution would be to enable and enforce branch protection rules. But yes, that would require input from @vbraun.

**Review by dimpase — APPROVED** [review]
lgtm. At worst, we break CI if it doesn't quite work.

**Comment by kwankyu** [comment]
All checks passed! 

Merging CI-fixes immediately to CI-workflows is a nice idea. Thanks!

File: .ci/merge-fixes.sh
Comment (tobiasdiez) [inline .ci/merge-fixes.sh:5]: You probably want to make sure that they also have the "postive review" label.

File: .github/workflows/build.yml
Comment (tobiasdiez) [inline .github/workflows/build.yml:27]: This adds now another row in the checks under each PR, which is already a pretty crowded. Can this not be merged directly in the build workflow as you do say with conda?

**Review by tobiasdiez — DISMISSED** [review]

File: .ci/merge-fixes.sh
Comment (mkoeppe) [reply .ci/merge-fixes.sh:5]: Definitely not. Sometimes it takes a while to find a competent reviewer.

File: .github/workflows/build.yml
Comment (mkoeppe) [reply .github/workflows/build.yml:27]: No, because the container does not have a suitable version of the gh script available, and installing it is too annoying.

File: .ci/merge-fixes.sh
Comment (kwankyu) [reply .ci/merge-fixes.sh:5]: See https://github.com/sagemath/sage/pull/36338#issuecomment-1736131315 for @mkoeppe's opinion on this.

Without requiring "positive review" label, we can check the PR before we set it positive review.

File: .github/workflows/build.yml
Comment (kwankyu) [reply .github/workflows/build.yml:27]: @mkoeppe, No response to this?

Comment (mkoeppe) [reply .github/workflows/build.yml:27]: You may need to switch a different view; I responded above

Comment (tobiasdiez) [reply .github/workflows/build.yml:27]: https://github.com/marketplace/actions/install-gh-cli or something similar?

File: .ci/merge-fixes.sh
Comment (tobiasdiez) [reply .ci/merge-fixes.sh:5]: It will influence every other PR, so I think there should be some safety measure.

> Without requiring "positive review" label, we can check the PR before we set it positive review.

What do you mean? Doesn't the CI of that PR already checks the PR? Why do you need other PRs to check it?

File: .github/workflows/build.yml
Comment (mkoeppe) [reply .github/workflows/build.yml:27]: that one's not portable to i686, and overall I don't think this is a worthwhile change

File: .ci/merge-fixes.sh
Comment (mkoeppe) [reply .ci/merge-fixes.sh:5]: > so I think there should be some safety measure.

We already have that, namely the Triage group + common sense.

Comment (kwankyu) [reply .ci/merge-fixes.sh:5]: > > Without requiring "positive review" label, we can check the PR before we set it positive review.
> 
> What do you mean? Doesn't the CI of that PR already checks the PR? 

I was confused and you are right.

Then I am inclined to agree with you. People with good intentions can make mistakes. Another safety measure does not hurt.

Comment (mkoeppe) [reply .ci/merge-fixes.sh:5]: Requiring a positive review does hurt because it slows down the response. It introduces another uncertainty for the potential reviewers: Does "positive review" now mean "good enough that it should take immediate effect because it's better than a completely broken CI", or does it mean "perfect with all bikeshedding discussions completed"?

Comment (tobiasdiez) [reply .ci/merge-fixes.sh:5]: Still better in my opinion than that the combo "blocker + needs review" would mean "everyone now has to cope with my experimental changes that nobody looked at yet". "blocker" is priority label (i.e. meaning: please have a look at this asap) and nothing more.

If you want to make the process easier for everyone, just prevent that PRs are merged without them passing the CI...

Comment (mkoeppe) [reply .ci/merge-fixes.sh:5]: In actual practice observed in our repo, people here simply do not use the "blocker" label frivolously. We can trust our developers.

And of course I have considered other possibilities before making the decision on "blocker", regardless of review status; in particular, introducing a separate label just for this task.

Comment (mkoeppe) [reply .ci/merge-fixes.sh:5]: Let's please just keep it simple. A practical problem, a simple solution.

Comment (kwankyu) [reply .ci/merge-fixes.sh:5]: > Requiring a positive review does hurt because it slows down the response.

We are not putting off fire. There's no need to act immediately as soon as someone pushes a CI-fix PR and adds blocker label.  We all want to fix CI failures as soon as possible and it would not take long before another one would lend second eyes to check the PR and give positive review.

But, as the author of this PR wants it that way, please @tobiasdiez waive your (and my) worries about broken Ci-fix PRs. If the worries turns out to be real, then we can fix it then (I guess @mkoeppe is the first to fix it :-)

File: .github/workflows/build.yml
Comment (kwankyu) [reply .github/workflows/build.yml:27]: I don't like the duplicate lines too. But that's a trivial issue, and there's no simple fix. Let's ignore it.

Comment (tobiasdiez) [reply .github/workflows/build.yml:27]: We are not using any i686 architecture...that's a very potential if and when argument...

File: .ci/merge-fixes.sh
Comment (tobiasdiez) [reply .ci/merge-fixes.sh:5]: Sorry, but I don't understand the argument here: because two people don't like the change, we go along with the solution that one person prefers?

I actually think this PR here is a perfect example: there is absolutely no reason this is tagged as a blocker, it doesn't resolve any issues with falling CIs or something similar, it is not tested very well (or did any of the reviewers tried to run the new macos workflow and the other changed workflows on the main repo?) and there are open discussions about how to implement certain things without too much disadvantages (it's another question how serious these points are). Why should this PR be merged into every other PR?

I'm happy with either of the following solutions:
- Introduce a new label that triggers the auto-merging (that is then assigned once someone looked over the changes and gave them a thumps up, even if there might be some smaller issues still outstanding)
- Auto-merge all positively reviewed PRs (that would be then a poor-mans merge queue)
- Only merge with positive review + blocker.

Comment (kwankyu) [reply .ci/merge-fixes.sh:5]: > Sorry, but I don't understand the argument here: because two people don't like the change, we go along with the solution that one person prefers?

I don't count, I weigh, and the opinion of the author (who did most of the work) weighs more to me.

**Comment by dimpase** [comment]
are we not reinventing the wheel here, by any chance? 
https://github.blog/changelog/2023-07-12-pull-request-merge-queue-is-now-generally-available/

File: .ci/merge-fixes.sh
Comment (mkoeppe) [reply .ci/merge-fixes.sh:5]: > I actually think this PR here is a perfect example: there is absolutely no reason this is tagged as a blocker

Actually there is. I marked it "blocker" so that the correct behavior of testing blockers is tested. 
This was to address a concern by @kliem in https://github.com/sagemath/sage/pull/36338#discussion_r1336694786

Comment (mkoeppe) [reply .ci/merge-fixes.sh:5]: > We are not putting off fire. There's no need to act immediately as soon as someone pushes a CI-fix PR and adds blocker label. 

Well, it's not fire, but it's a matter of "availability of the CI services", which, if poor, leads to developer frustration (@fchapoton @tornaria) and can dramatically reduce development velocity because developers may disengage or wait until the CI is fixed again.

Let's consider the period from the release of a beta version to the release of the next beta version, currently about a week. For, say, 95% of availability of the CI, we want a CI fix to take effect within a few hours of a broken release.

> We all want to fix CI failures as soon as possible and it would not take long before another one would lend second eyes to check the PR and give positive review.

This may be true for trivial fixes such as the linter failures, which can be fixed and reviewed by many people.

But for fixes to CI and build system, it hinges on the availability of a very small number of people, and time-to-positive-review can be several days -- which can easily make the availability of the CI services less than 50%.

File: .github/workflows/build.yml
Comment (mkoeppe) [reply .github/workflows/build.yml:27]: > We are not using any i686 architecture...that's a very potential if and when argument...

Yes, we are. The build.yml workflow accepts the container image on workflow_dispatch, and we do test 32-bit platforms.

**Comment by mkoeppe** [comment]
> are we not reinventing the wheel here, by any chance?
> https://github.blog/changelog/2023-07-12-pull-request-merge-queue-is-now-generally-available/

Yes, @dimpase, I have considered many options, including Merge Queues. But moving to using Merge Queues is not actionable for the project; the release manager would have to adopt a changed workflow. As I wrote in https://github.com/sagemath/sage/pull/36338#issuecomment-1736027497:

> Yes, this PR is a practical solution. It is designed to avoid two approaches, each of which would be more elegant:
> 
> * It does not create hotfix branches explicitly. Hence, there is no hotfix branch to maintain; hence there is not need for a maintainer of the hotfix branch.
> * It does not require communication with the release manager.
> 
> This careful design makes this PR an effective solution.

File: .ci/merge-fixes.sh
Comment (mkoeppe) [reply .ci/merge-fixes.sh:5]: @tobiasdiez wrote:
> Sorry, but I don't understand the argument here: because two people don't like the change, we go along with the solution that one person prefers?

I know that @tobiasdiez knows this already, but I'll point out again: Our project does not hold majority votes In PR discussions.

> I actually think this PR here is a perfect example: there is absolutely no reason this is tagged as a blocker, it doesn't resolve any issues with falling CIs or something similar, 

This is false. This PR does resolve the issue of the failing CIs. Merging this PR is enough to fix all CI problems in the current cycle: it auto-applies the blocker PRs #36327 and #36276.

> it is not tested very well (or did any of the reviewers tried to run the new macos workflow and the other changed workflows on the main repo?)

This is false, and moreover an inappropriate attack on author and reviewers.

Comment (kwankyu) [reply .ci/merge-fixes.sh:5]: OK. Here is my suggestion. As we can observe by looking at the labelling war below, there are conflicting opinions on which way this PR should proceed. If the issue is large and of broad interest in our community, then this situation would be resolved by a voting in sage-devel. But the current issue is a small one, and it would be ridiculous and shameful to spend a long time for endless discussions or a large-scale voting. Hence let me suggest a *small voting* here, as defined below:

The voting is about (a) "only blocker label" vs (b) "blocker label + positive review". The voting occurs only in this thread. The voting starts by the votes of @tobiasdiez and @mkoeppe, by which they are supposed to have vowed on accepting the voting result. The voting ends automatically when there was no more vote for the last 24 hours. When the voting ends, the author of the PR proceeds in the way of winning candidate. Please only vote with an optional simple comment in the rest of this thread.

Comment (mkoeppe) [reply .ci/merge-fixes.sh:5]: -1 on holding majority votes in PRs.

**Comment by dimpase** [comment]
I don't see how a merge queue with blockers (or positively reviewed blockers) is different in functionality from this PR.

**Comment by mkoeppe** [comment]
> I don't see how a merge queue with blockers (or positively reviewed blockers) is different in functionality from this PR.

Well, Merge Queues are attached to a branch protection rule for a branch. A PR can only be merged in one branch, namely its target branch. So doing anything like this would interfere with the release manager's work.

**Comment by mkoeppe** [comment]
@dimpase You may want to try out merge queues with https://github.com/dimpase/primecountpy/pulls ...

**Comment by dimpase** [comment]
> @dimpase You may want to try out merge queues with https://github.com/dimpase/primecountpy/pulls ...

I don't think it's a repo which needs them.

**Comment by mkoeppe** [comment]
But it's harmless to play with them there.

**Comment by kwankyu** [comment]
Thanks for the benevolent action, @vbraun!

File: .ci/merge-fixes.sh
Comment (tobiasdiez) [reply .ci/merge-fixes.sh:5]: > @tobiasdiez wrote:
> 
> > Sorry, but I don't understand the argument here: because two people don't like the change, we go along with the solution that one person prefers?
> 
> I know that @tobiasdiez knows this already, but I'll point out again: Our project does not hold majority votes In PR discussions.
> 
> > I actually think this PR here is a perfect example: there is absolutely no reason this is tagged as a blocker, it doesn't resolve any issues with falling CIs or something similar,
> 
> This is false. This PR does resolve the issue of the failing CIs. Merging this PR is enough to fix all CI problems in the current cycle: it auto-applies the blocker PRs #36327 and #36276.
> 
> > it is not tested very well (or did any of the reviewers tried to run the new macos workflow and the other changed workflows on the main repo?)
> 
> This is false, and moreover an inappropriate attack on author and reviewers.

Sadly not false...

**Pull request merged**
